### ISSUE TITLE
TLS with CA signed certificate

### ISSUE THREAD

**Issue by MikalFurnell** [neutrinolabs/xrdp#2449]
Hello,

I am trying to get XRDP working with Apache Guacamole and can't seem to get it to work with the TLS certificate in place.

Here is the applicable xrdp.ini configuration(let me know if you need more):
```
security_layer=tls
certificate=
key_file=
ssl_protocols=TLSv1.2, TLSv1.3
tls_ciphers=HIGH
```

Here is the certificate:
```
root@...:/etc/xrdp# openssl x509 -in cert.pem -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            28:00:00:00:15:c9:22:37:7e:37:95:03:32:00:00:00:00:00:15
        Signature Algorithm: sha512WithRSAEncryption
        Issuer: DC = ..., DC = ..., CN = ...
        Validity
            Not Before: Dec  2 19:53:35 2022 GMT
            Not After : Dec  2 19:53:35 2023 GMT
        Subject: CN = ...
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    00:c9:f4:00:9d:13:36:39:d0:43:0f:7e:82:5f:3f:
                    44:5c:9c:c8:3c:ab:fe:24:1e:e7:4b:bd:1b:aa:1c:
                    b9:13:1a:00:2a:db:b4:42:0d:2b:1f:f7:a5:94:89:
                    42:76:0b:4f:ce:d5:76:e2:20:b5:06:4f:28:67:db:
                    ec:ce:56:e8:5b:fe:47:37:09:a8:38:41:4d:a9:08:
                    7b:d5:27:b2:7f:c6:0b:09:4c:38:0a:79:9b:5b:b1:
                    f2:f4:c2:8b:17:d7:79:d1:f5:1b:6e:b5:db:40:0e:
                    f8:9c:8b:3f:e8:8b:3d:a0:aa:e7:70:c4:da:b6:49:
                    c8:e4:43:c7:5d:74:20:af:d5:a8:ea:5b:ec:c4:5b:
                    98:ef:c6:cb:37:7c:53:85:1d:18:51:f3:2a:38:7a:
                    09:b6:76:a2:31:55:5b:65:fc:b2:61:88:40:a6:4a:
                    66:73:0c:02:2a:5c:e7:98:19:6f:33:d8:94:5c:f1:
                    67:1d:e2:5a:36:4f:bb:b8:c0:a0:62:97:79:cd:8d:
                    5f:73:54:b4:89:10:86:40:4d:25:17:8c:58:5d:d0:
                    c4:84:f3:0c:bf:89:7b:4b:5a:4a:f0:d8:27:74:af:
                    fa:6f:97:90:b8:f7:43:e2:82:ab:fe:0e:f9:8d:79:
                    37:e5:ae:52:13:7c:8f:65:c1:99:73:dc:32:55:de:
                    36:dd
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            1.3.6.1.4.1.311.21.7:
                0-.%+.....7.................9....#...]..b..d...
            X509v3 Extended Key Usage:
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            1.3.6.1.4.1.311.21.10:
                0.0
..+.......0
..+.......
            X509v3 Subject Key Identifier:
                2D:73:66:B1:06:51:C6:86:A7:9F:81:ED:73:C8:F4:B2:23:08:75:DA
            X509v3 Subject Alternative Name:
                DNS:..., DNS:..., IP Address:192.168.20.13
            X509v3 Authority Key Identifier:
                keyid:1D:79:F6:28:E3:1A:23:3C:BE:2D:2C:84:F6:4C:D3:6D:E2:F6:99:5B

            X509v3 CRL Distribution Points:

                Full Name:
                  URI:ldap:///...

            Authority Information Access:
                CA Issuers - URI:ldap:///...

    Signature Algorithm: sha512WithRSAEncryption
         8c:bf:a1:ea:7b:28:9d:67:49:3a:0c:57:8d:ba:d9:37:85:fc:
         95:99:52:e2:c9:65:56:a3:5a:25:d3:01:37:53:01:0a:5a:04:
         c1:08:c1:66:2c:ae:e1:7e:ff:5d:0d:fe:36:91:ae:8d:65:3d:
         5f:5f:42:8f:91:87:b8:75:de:0a:7a:23:56:28:81:57:c4:ef:
         4f:d5:31:96:40:ab:97:14:a3:0d:5a:70:25:57:7c:89:c1:79:
         36:bf:f8:0e:4c:29:02:f4:9a:50:e4:60:b1:99:2a:6f:20:f3:
         3e:d8:61:aa:fd:27:72:50:4d:19:f3:58:05:e5:ec:dc:3d:5e:
         4c:ec:12:41:24:b6:61:6a:36:e5:89:c1:b6:29:f7:ad:94:7e:
         6b:2a:ed:6c:39:d7:00:c7:bc:81:eb:fb:46:86:30:f1:60:71:
         d3:68:7f:5d:98:7d:c4:bd:a4:4f:b1:1e:c3:fc:70:b9:a0:bc:
         25:cb:8b:2b:0b:96:fb:b4:72:b2:df:cb:11:bd:60:73:71:d7:
         83:a3:a7:43:80:f8:2d:a3:6e:d2:32:41:a8:74:0e:be:70:d2:
         80:79:09:68:2c:a7:32:78:14:24:3a:1e:57:d7:62:49:99:27:
         33:bf:d6:85:34:14:6b:96:a9:7a:88:5f:41:65:7b:f5:17:63:
         82:74:70:98
```

Here is what I get when I try to connect using Guacamole:
```
[20221205-16:19:05] [INFO ] starting xrdp with pid 2212
[20221205-16:19:05] [INFO ] address [0.0.0.0] port [3389] mode 1
[20221205-16:19:05] [INFO ] listening to port 3389 on 0.0.0.0
[20221205-16:19:05] [INFO ] xrdp_listen_pp done
[20221205-16:19:09] [INFO ] Socket 12: AF_INET6 connection received from ::ffff:... port 36216
[20221205-16:19:09] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:... port 3389)
[20221205-16:19:09] [DEBUG] Closed socket 11 (AF_INET6 :: port 3389)
[20221205-16:19:09] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20221205-16:19:09] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20221205-16:19:09] [DEBUG] TLSv1.3 enabled
[20221205-16:19:09] [DEBUG] TLSv1.2 enabled
[20221205-16:19:09] [DEBUG] Security layer: requested 1, selected 1
ssl_tls_accept: TLS connection accepted
libxrdp_force_read: error
xrdp_process_main_loop: libxrdp_process_incoming failed
got XRDP SIGPIPE(13)
got XRDP SIGPIPE(13)
ssl_tls_print_error: SSL_shutdown: Server closed TLS connection
[20221205-16:19:09] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:... port 3389)
xrdp_mcs_disconnect - socket closed
```

And here is what I get when I try to just look and see if I can see the certificate on the port:
```
root@...:/etc/xrdp# openssl s_client -connect ...:3389
CONNECTED(00000003)
140630150374720:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:331:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 5 bytes and written 283 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

**Comment by J-W-C** [comment]
Seems like a client-side config issue.

Would not expect the openssl s_client test to succeed.  RDP servers don't answer port 3389 with an immediate TLS handshake, they need to get the RDP equivalent of STARTTLS

Maybe bring it up on Gitter, https://gitter.im/neutrinolabs/xrdp-questions

**Comment by matt335672** [comment]
Indeed - the s_client won't work with an RDP server in a standard configuration.

It looks to me like the client is going away. Some ideas for you:-
- Guacamole can't validate the certificate and unlike a regular RDP client can't prompt the user for what to do. Here's [an example](https://stackoverflow.com/questions/68458381/guacamole-rdp-wont-connect) [stackoverflow.com]
- Try using a connection from `xfreerdp` on the Guacamole machine and see if you get any client-side errors.

**Comment by MikalFurnell** [comment]
Appreciate the responses. I think maybe part of what put me on that path was I have a windows machine with RDP configured with a certificate on it and it presents the certificate when utilizing openssl. Also the only thing that I change from a working to a non-working connection is the check box for the connection that allows it to ignore the server certificate.

I spent some time looking into it and it does look like a client issue. I copied the CA certificate into the guacd container and was able to test using openssl and it doesn't trust the CA certificate. Ill have to work on adding that to the container and then everything will probably work. Thank you for the help!
